### ISSUE TITLE
Assignment.due_at wrong when there are overrides

### ISSUE THREAD

**Issue by fredrikekre** [instructure/canvas-lms#1547]
## Summary:

When there are overrides for an assignment `Assignment.due_at` reports the latest time of all overrides contrary to what the [documentation](https://canvas.instructure.com/doc/api/assignments#Assignment) says:

>   // the due date for the assignment. returns null if not present. NOTE: If this
  // assignment has assignment overrides, this field will be the due date as it
  // applies to the user requesting information from the API.

### Steps to reproduce:

 1. Create an assignment
 2. Add an override for a student with later due date
 3. Request the assignment from the `GET /api/v1/courses/:course_id/assignments/:id` endpoint

### Additional notes:
See also this old post https://community.canvaslms.com/thread/7269

**Comment by Ardena** [comment]
fredrikekre - Are you requesting the due date as a teacher or admin?  The due dates are calculated as most lenient, so when teachers and admins request the information, they get the latest due date since they can see all of the due dates.  You can also get the overrides from the API on their own, which, I believe, includes the "everyone" due date if it exists.

**Comment by fredrikekre** [comment]
As a teacher.

>  The due dates are calculated as most lenient, so when teachers and admins request the information, they get the latest due date [...]

I would argue that the "best" date to put in `due_date` is the original date. When you ask "When is this assignment due?" (which I would argue `Assignment.due_at` is intended to answer) you probably mean/expect the base date. But if this is indeed working as intended it would be good to update the documentation.

> [...] since they can see all of the due dates.

Right, so I can get this by requesting `"include[]": "all_dates"` (or `"all_dates": "true"` when querying a single assignment) and look for `"base": true` to find the base date, but it just feels like jumping through unecessary hoops to get this information. However, as mentioned in https://community.canvaslms.com/thread/7269 you can not get this information when you query submissions with `"include[]": "assignment"` for example.

Also, since teachers can get all dates it makes, IMO, sense to put the base date in `due_at` and if you are interested in the other dates you use `"include[]": "all_dates"`/`"all_dates": "true"`.

**Comment by Ardena** [comment]
I agree with you about "best" due date, since you have the option to view the overrides, but it may be based on when the teacher should expect all the submissions to be turned in and they can start grading.

This is something I've wondered about occasionally as well, so I've added it to our list to look into.

**Comment by GideonWolfe** [comment]
Is there any progress on this issue? I am a student trying to write a program to check on assignments, and I am simply unable to access a correct due date in any format for any API call.

**Comment by simonista** [comment]
@GideonWolfe : I think your issue is different than the one reported. You should be getting the correct due_at value from the api as your student user. The previous reported issue was about teachers. Can you say more about what you are seeing?

**Comment by GideonWolfe** [comment]
@simonista sure! 

We are pulling down a list of assignments and the dates for `due_at` do not reflect the due dates we see on our course pages.

It seems like we noticed a pattern that all reported `due_at` dates are 19 hours off. For example, my professor has every single assignment due at `11:59 PM`. But the `due_at` value for every assignment is for `6:59 PM` the following day.

We could account for this in our code since there seems to be a pattern, but I don't know if it will always hold and I would rather it just give me the correct value. I can start a new issue if you would like me to provide specifics. The code I am writing is [here])https://github.com/GideonWolfe/canvas-cli).

**Comment by simonista** [comment]
Are you accounting for timzone offsets? timestamps are always returned in UTC in the api, and 6:59 actually sounds potentially like 6:59 AM, which would be 7 hours difference, which is what you'd be seeing if you're in the pacific time zone.

**Comment by GideonWolfe** [comment]
Ah that makes sense! I assumed that the timezone returned corresponded to the timezone defined within the course.

**Comment by cycomachead** [comment]
I would like to know if there's any intent on addressing this. We run into this issue when fetching all assignments at once. There are newer APIs that require additional calls per-assignment. 

It would be helpful if as an instructor / staff member the `due_date` was always the original one. There's really not much in the API response to indicate whether the API response would return inconsistent data.
